### PR TITLE
ref(ddm): streamline MRI naming and parsing

### PR DESCRIPTION
**Pull request by obostjancic**
- Implements https://www.notion.so/sentry/DDM-Frontend-Naming-e58cba9bbd444f67aa387d4dd70058bc
- Re-implements MRI parsing, Introduces MRI type
- Implements MRI field parsing and formatting

### CODE REVIEW

File: static/app/types/metrics.tsx [hunk 1]
```diff
@@ -13,6 +11,19 @@ export type MetricsOperation =
...
+
+export type MRI = `${MetricType}:${UseCase}${string}@${string}`;
+
+export type ParsedMRI = {
+  name: string;
+  type: MetricType;
+  unit: string;
+  useCase: UseCase;
+};
+
```
Comment (ArthurKnaus) [inline static/app/types/metrics.tsx:26]: love it ❤️

File: static/app/utils/metrics/mri.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+export function MRIToField(mri: MRI, op: string): string {
+  return `${op}(${mri})`;
+}
+
+export function parseField(field: string): {mri: MRI; op: string} | null {
+  const parsedFunction = parseFunction(field);
+  if (!parsedFunction) {
+    // We only allow aggregate functions for custom metric alerts
```
Comment (ArthurKnaus) [inline static/app/utils/metrics/mri.tsx:65]: Probably not needed anymore in this context?

File: static/app/utils/metrics/mri.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+export function formatMRI(mri: MRI): string {
+  return parseMRI(mri)?.name ?? mri;
+}
+
+export function getUseCaseFromMRI(mri?: string): UseCase {
+  const parsed = parseMRI(mri);
+
+  return parsed?.useCase ?? 'sessions';
```
Comment (ArthurKnaus) [inline static/app/utils/metrics/mri.tsx:55]: Why is the default `sessions`? 
Shouldn't we just return `null` here and let the default value be the concern of the code that uses it?

File: static/app/utils/metrics/mri.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  };
+}
+
+// convenience function to get the MRI from a field, returns defaut MRI if it fails
+export function getMRI(field: string): MRI {
+  const parsed = parseField(field);
+  return parsed?.mri ?? DEFAULT_MRI;
+}
+
+export function formatMRIAggregate(aggregate: string) {
```
Comment (ArthurKnaus) [inline static/app/utils/metrics/mri.tsx:80]:
```suggestion
export function formatMRIField(aggregate: string) {
```

File: static/app/utils/metrics/mri.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
+import * as Sentry from '@sentry/react';
+
+import {t} from 'sentry/locale';
+import {MetricType, MRI, ParsedMRI} from 'sentry/types/metrics';
+import {parseFunction} from 'sentry/utils/discover/fields';
+
+import {DEFAULT_METRIC_ALERT_AGGREGATE, UseCase} from './index';
```
Comment (ArthurKnaus) [inline static/app/utils/metrics/mri.tsx:7]: should we move `DEFAULT_METRIC_ALERT_AGGREGATE` in here and rename it to `DEFAULT_METRIC_ALERT_FIELD`?

File: static/app/views/alerts/rules/metric/utils/isCustomMetricAggregate.tsx [hunk 6]
```diff
@@ -1,6 +1,6 @@
-import {fieldToMri, getUseCaseFromMRI} from 'sentry/utils/metrics';
+import {getUseCaseFromMRI, parseField} from 'sentry/utils/metrics/mri';
 
 export const isCustomMetricAggregate = (aggregate: string): boolean => {
```
Comment (ArthurKnaus) [inline static/app/views/alerts/rules/metric/utils/isCustomMetricAggregate.tsx:3]:
```suggestion
export const isCustomMetricField = (aggregate: string): boolean => {
```

File: static/app/views/alerts/rules/metric/utils/isCustomMetricAlert.tsx [hunk 7]
```diff
@@ -1,8 +1,8 @@
-import {fieldToMri} from 'sentry/utils/metrics';
+import {parseField} from 'sentry/utils/metrics/mri';
 
 /**
  * Currently we can tell if an alert is a crash free alert by checking the aggregate for a MRI,
  */
 export function isCustomMetricAlert(aggregate: string): boolean {
-  return fieldToMri(aggregate).mri !== undefined;
+  return Boolean(parseField(aggregate)?.mri);
```
Comment (ArthurKnaus) [inline static/app/views/alerts/rules/metric/utils/isCustomMetricAlert.tsx:7]: nit: we prefer `!!`
```suggestion
  return !!parseField(aggregate)?.mri;
```

**Review by ArthurKnaus — APPROVED** [review]
Nice work! 🎉

File: static/app/types/metrics.tsx [hunk 8]
```diff
@@ -59,12 +70,13 @@ export type MetricsTagValue = {
   value: string;
 };
 
-export type MetricsMeta = {
-  mri: string;
-  name: string;
+export type MetricMeta = {
+  mri: MRI;
+  // name is returned by the API but should not be used, use parseMRI(mri).name instead
+  // name: string;
```
Comment (matejminar) [inline static/app/types/metrics.tsx:76]: Useful comment ❤️

File: static/app/types/metrics.tsx [hunk 9]
```diff
@@ -13,6 +11,19 @@ export type MetricsOperation =
   | 'p95'
   | 'p99';
 
+export type MetricType = 'c' | 'd' | 'g' | 'e' | 's';
+
+export type UseCase = 'custom' | 'transactions' | 'sessions';
+
+export type MRI = `${MetricType}:${UseCase}${string}@${string}`;
```
Comment (matejminar) [inline static/app/types/metrics.tsx:18]: 🚀
I wonder what's the perf of this. Fingers crossed.

**Comment by obostjancic** [comment]
@ArthurKnaus I think I addressed all of the comments in [8208e52](https://github.com/getsentry/sentry/pull/60543/commits/8208e525b9167c44e5e537e8643863ec40b1abb8)

**Review by matejminar — APPROVED** [review]

**Pull request merged**